File: chap04/Sorting/Iterative_Quick_Sort1.cpp
// CPP code for recursive function of Quicksort
#include <bits/stdc++.h>
using namespace std;

/* This function takes last element as pivot, places the pivot element
   at its correct position in sorted array, and places all smaller
   (smaller than pivot) to left of pivot and all greater elements to
   right of pivot */
int partition( int arr[], int l, int h) {
  int x = arr[h];
  int i = (l-1);

  for( int j=l; j<=h-1; j++) {
    if( arr[j]<=x) {
      i++;
      swap(arr[i], arr[j]);
    }
  }
  swap( arr[i+1], arr[h]);
  return (i+1);
}

/* A[] --> Array to be sorted, l --> Starting index,
   h --> Ending index */
void quickSort( int A[], int l, int h)
{
  if( l<h)  {
    // Partitioning index
    int p = partition( A, l, h);
    quickSort( A, l, p-1);
    quickSort(A, p+1, h);
  }
}

int main()
{
  int n=5;
  int arr[n] = { 4, 2, 6, 9, 2 };

  quickSort( arr, 0, n-1);

  for( int i=0; i<n; i++) {
    cout << arr[i] << " ";
  }

  return 0;
}

File: chap04/Sorting/Iterative_Quick_Sort1.java
// Java program for implementation of QuickSort
import java.util.*;

class QuickSort {
  /* This function takes last element as pivot, places the pivot element
  at its correct position in sorted array, and places all smaller
  (smaller than pivot) to left of pivot and all greater elements to right
  of pivot */
  static int parition( int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low-1); // index of smaller element

    for( int j=low; j<=high-1; j++) {
      // If current element is smaller than or equal to povot
      if( arr[j] <= pivot) {
        i++;

        // swap arr[i] and arr[j]
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
      }
    }

    // swap arr[i+1] and arr[high] (or pivot)
    int temp = arr[ i+1];
    arr[ i+1] = arr[ high];
    arr[ high] = temp;

    return i+1;
  }

  /* The main function that implements QuickSort() 
  arr[] --> Array to be sorted,
  low --> Starting index,
  high --> Ending index */
  static void qSort( int arr[], int low, int high) {
    if( low<high) {
      // pi is partitioning index, arr[pi] is now at right place
      int pi = partition( arr, low, high);

      // Recursively sort elements befere partition
      // and right place
      qSort( arr, low, pi-1);
      qSort( arr, pi+1, high);
    }
  }
}
    
